### ISSUE TITLE
Error on compilation: "Cannot find module '@gi/Shell'"

### ISSUE THREAD

**Issue by kcpr** [yilozt/rounded-window-corners#62]
Hi.

While trying to compile the source code and install extension (with `yarn install && yarn run ext:install`) I experience the following errors:
```node
yarn install v1.22.19
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.64s.
yarn run v1.22.19
$ gulp install
[01:43:22] Using gulpfile ~/Projects/rounded-window-corners/gulpfile.js
[01:43:22] Starting 'install'...
[01:43:22] Starting '<anonymous>'...
[01:43:22] Finished '<anonymous>' after 2.73 ms
[01:43:22] Starting 'generate_gi_ext'...
[01:43:22] Starting 'generate_gi_prefs'...
[01:43:22] Finished 'generate_gi_ext' after 1.16 ms
[01:43:22] Finished 'generate_gi_prefs' after 1.72 ms
[01:43:22] Starting 'rename_gi'...
[01:43:22] Finished 'rename_gi' after 9.82 ms
[01:43:22] Starting '<anonymous>'...
[01:43:22] Finished '<anonymous>' after 1.25 ms
[01:43:22] Starting 'format'...
[01:43:28] Finished 'format' after 6.25 s
[01:43:28] Starting 'compile_ts'...
src/effect/clip_shadow_effect.ts(4,41): error TS2792: Cannot find module '@gi/Shell'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
src/effect/clip_shadow_effect.ts(24,18): error TS2339: Property 'add_glsl_snippet' does not exist on type '(Anonymous class)'.
src/effect/clip_shadow_effect.ts(29,18): error TS2339: Property 'get_pipeline' does not exist on type '(Anonymous class)'.
src/effect/linear_filter_effect.ts(5,41): error TS2792: Cannot find module '@gi/Shell'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
src/effect/linear_filter_effect.ts(11,18): error TS2339: Property 'add_glsl_snippet' does not exist on type '(Anonymous class)'.
src/effect/linear_filter_effect.ts(15,18): error TS2339: Property 'get_pipeline' does not exist on type '(Anonymous class)'.
src/effect/rounded_corners_effect.ts(4,41): error TS2792: Cannot find module '@gi/Shell'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
src/effect/rounded_corners_effect.ts(60,26): error TS2339: Property 'get_uniform_location' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(66,18): error TS2339: Property 'add_glsl_snippet' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(84,32): error TS2339: Property 'actor' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(114,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(115,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(116,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(117,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(118,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(119,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(120,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(121,18): error TS2339: Property 'set_uniform_float' does not exist on type 'Effect'.
src/effect/rounded_corners_effect.ts(124,18): error TS2339: Property 'queue_repaint' does not exist on type 'Effect'.
src/extension.ts(29,28): error TS2792: Cannot find module '@gi/Shell'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
src/extension.ts(174,43): error TS2339: Property 'window_container' does not exist on type 'WindowPreview'.
src/extension.ts(178,38): error TS2345: Argument of type 'RegisteredPrototype<(Anonymous class), { [key: string]: ParamSpec<unknown>; }, unknown[]>' is not assignable to parameter of type 'Effect'.
  Type 'Properties<(Anonymous class), SnakeToCamel<{ [key: string]: ParamSpec<unknown>; }> & SnakeToUnderscore<{ [key: string]: ParamSpec<unknown>; }>> & (Anonymous class)' is missing the following properties from type 'Effect': _init, queue_repaint, vfunc_modify_paint_volume, vfunc_paint, and 51 more.
src/extension.ts(185,18): error TS2339: Property 'insert_child_below' does not exist on type 'WindowPreview'.
src/extension.ts(189,24): error TS2769: No overload matches this call.
  Overload 1 of 2, '(source: Object, signal: "notify", callback: (_source: Object, pspec: ParamSpec<unknown>) => void): void', gave the following error.
    Argument of type 'WindowPreview' is not assignable to parameter of type 'Object'.
      Type 'WindowPreview' is missing the following properties from type 'Object': _init, g_type_instance, connect, connect_after, and 34 more.
  Overload 2 of 2, '(source: Object, args_0: string, args_1: (_: unknown[]) => unknown): void', gave the following error.
    Argument of type 'WindowPreview' is not assignable to parameter of type 'Object'.
src/extension.ts(192,35): error TS2345: Argument of type 'WindowPreview' is not assignable to parameter of type 'Object'.
src/extension.ts(397,40): error TS2339: Property 'window_container' does not exist on type 'WindowPreview'.
src/extension.ts(398,40): error TS2339: Property 'get_allocation_box' does not exist on type 'WindowPreview'.
src/manager/rounded_corners_manager.ts(23,28): error TS2792: Cannot find module '@gi/Shell'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
src/manager/rounded_corners_manager.ts(179,51): error TS7006: Parameter 'actor' implicitly has an 'any' type.
src/manager/rounded_corners_manager.ts(198,24): error TS7006: Parameter 'actor' implicitly has an 'any' type.
src/manager/rounded_corners_manager.ts(305,13): error TS2345: Argument of type 'RegisteredPrototype<(Anonymous class), { [key: string]: ParamSpec<unknown>; }, unknown[]>' is not assignable to parameter of type 'Effect'.
  Type 'Properties<(Anonymous class), SnakeToCamel<{ [key: string]: ParamSpec<unknown>; }> & SnakeToUnderscore<{ [key: string]: ParamSpec<unknown>; }>> & (Anonymous class)' is missing the following properties from type 'Effect': _init, queue_repaint, vfunc_modify_paint_volume, vfunc_paint, and 51 more.
src/manager/rounded_corners_manager.ts(431,61): error TS2345: Argument of type 'RegisteredPrototype<Effect, { [key: string]: ParamSpec<unknown>; }, unknown[]>' is not assignable to parameter of type 'Effect'.
src/manager/rounded_corners_manager.ts(651,47): error TS7006: Parameter 'actor' implicitly has an 'any' type.
src/manager/rounded_corners_manager.ts(776,58): error TS2345: Argument of type 'RegisteredPrototype<Effect, { [key: string]: ParamSpec<unknown>; }, unknown[]>' is not assignable to parameter of type 'Effect'.
  Type 'Properties<Effect, SnakeToCamel<{ [key: string]: ParamSpec<unknown>; }> & SnakeToUnderscore<{ [key: string]: ParamSpec<unknown>; }>> & Effect' is missing the following properties from type 'Effect': _init, queue_repaint, vfunc_modify_paint_volume, vfunc_paint, and 51 more.
TypeScript: 34 semantic errors
TypeScript: emit succeeded (with errors)
[01:43:36] 'compile_ts' errored after 8.02 s
[01:43:36] Error: TypeScript: Compilation failed
    at Output.mightFinish (/home/k/Projects/rounded-window-corners/node_modules/gulp-typescript/release/output.js:131:43)
    at /home/k/Projects/rounded-window-corners/node_modules/gulp-typescript/release/output.js:44:22
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
[01:43:36] 'install' errored after 14 s
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
What should I do to resolve the issue?

**Comment by yilozt** [comment]
Can you provide your system informatiom, such as linux distribution and gnome shell version? I will try to reproduce this issue.

**Comment by yilozt** [comment]
Seems like the `@gi` folder has not been generated correctly, this folder contain the type define files for this extensions:

```
@gi
├── Adw.d.ts
├── Clutter.d.ts
├── Cogl.d.ts
├── Gdk.d.ts
├── Gio.d.ts
├── GIRepository.d.ts
├── GLib.d.ts
├── GObject.d.ts
├── Graphene.d.ts
├── Gtk.d.ts
├── Meta.d.ts
├── Shell.d.ts
└── St.d.ts
```

You can remove `package-lock.json`  `@gi`  `docs.lock` then generate this folder again:

```bash
rm -rf @gi package-lock.json docs.lock
yarn gi
```

The out put will like this:

```
yarn run v1.22.18
$ gulp gi
[09:35:33] Using gulpfile ~/code/gnome-shell-extensions-rounded-window/gulpfile.js
[09:35:33] Starting 'gi'...
[09:35:33] Starting '<anonymous>'...
[09:35:33] Finished '<anonymous>' after 2.03 ms
[09:35:33] Starting 'generate_gi_ext'...
[09:35:33] Starting 'generate_gi_prefs'...
Generate .d.ts files for preferences pages: 
Loading docs.json...
Finding GIR files...
Resolving libraries...
Loading GIR files...
Adding generics...
Injecting types...
Generated!

[09:35:47] Finished 'generate_gi_prefs' after 14 s
Generate .d.ts files for extensions: 
Loading docs.json...
Finding GIR files...
Resolving libraries...
Loading GIR files...
Adding generics...
Unknown type:  {
    "name": "gpu",
    "transfer-ownership": "none"
} 
Marking as unknown!
Unknown type:  {
    "name": "logical_monitor",
    "transfer-ownership": "none"
} 
Marking as unknown!
Injecting types...
>> Conflict in Text.position with Actor.position.
>> Conflict in Text.position with Actor.position.
Generated!

The following types were prefixed with __ to preserve valid JavaScript identifiers.
NM.__80211Mode = NM.80211Mode
NM.__80211ApFlags = NM.80211ApFlags
NM.__80211ApSecurityFlags = NM.80211ApSecurityFlags

[09:35:48] Finished 'generate_gi_ext' after 16 s
[09:35:48] Starting 'rename_gi'...
[09:35:48] Finished 'rename_gi' after 179 ms
[09:35:48] Starting '<anonymous>'...
[09:35:48] Finished '<anonymous>' after 2.49 ms
[09:35:48] Finished 'gi' after 16 s
Done in 26.56s.
```

**Comment by kcpr** [comment]
Thank You for the answers.

> Can you provide your system informatiom, such as linux distribution and gnome shell version? I will try to reproduce this issue.

It's Ubuntu 22.04, GNOME Shell 42.4. Node 16.16.0 - LTS, but tried other versions too as I use NVM.

I tried the following, but it didn't seem to help.
```
rm -rf @gi package-lock.json docs.lock
yarn gi
```

I noticed this error though:
```
Error: Failed to import PolkitAgent in Shell, not installed or accessible.
```
Can provide some more logs later if required.

**Comment by yilozt** [comment]
I have add build dependencies for building in README: https://github.com/yilozt/rounded-window-corners/#ubuntu-2204

I built extensions after install those packages:

```
apt install gnome-shell-common libpolkit-agent-1-dev libmutter-10-dev \
    libadwaita-1-dev libgtk-3-dev libgtk-4-dev libgraphene-1.0-dev    \
    libgirepository1.0-dev libnm-dev libgcr-3-dev gettext
```

You can remove `@gi` folder then run `yarn ext:install` to built it.

**Comment by kcpr** [comment]
Ok, I installed the packages and it seems that it worked this time. Thank You!
I've been missiong "libgck-1-dev", "libgcr-3-dev", "libp11-kit-dev" and "libpolkit-agent-1-dev" I think, but I've installed some packages manually before.

I like that You included this list in the README file - I believe others may find it useful too.

I believe the issue may be closed. Thank You again for the help and for creating the extension itself!
